<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <meta name="author" content="Data capentry contributors" />
  <title>examining SRA run table</title>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
  </style>
</head>
<body>
<div id="header">
<h1 class="title">examining SRA run table</h1>
<h2 class="author">Data capentry contributors</h2>
</div>
<h2 id="learning-objectives">Learning Objectives</h2>
<ul>
<li>Identify 3 ways to download data onto a server</li>
<li>Get a general understanding of the NCBI SRA</li>
<li>Understand how to download a summary run table of the SRA</li>
<li>Be able to open a run table using a spreadsheet editor</li>
<li>Be able to examine a run table (text file) at the command line</li>
</ul>
<h2 id="lesson">Lesson</h2>
<p>There are what we call &quot;command line tools&quot; that can download data from typical data sources such as FTP, HTTP, or HTTPS. Additionally, there are &quot;programs&quot; that are host-specific or database-specific, such as NCBI SRA tools (demostrated below.) Mainly, when you are downloading data you will be using one of the three options.</p>
<p>The first of the the commands to learn is <a href="https://en.wikipedia.org/wiki/Wget">wget</a> and <a href="https://en.wikipedia.org/wiki/Curl_(programming_language)">curl</a>. Both are programs that retrieve content from web servers. And both can be used effectively -- in other words, user-choice, for more details see <a href="http://daniel.haxx.se/docs/curl-vs-wget.html">this post</a>.</p>
<p>For example, if we wanted to download data from an FTP site, we could open a browser, and type in the the address. The result would be a sequence file, named REL606.fa.gz would download onto your computer.</p>
<p>Alternately, you could use download programs using wget.</p>
<pre><code>wget http://athyra.idyll.org/~t/REL606.fa.gz</code></pre>
<p>Or</p>
<pre><code>curl -O http://athyra.idyll.org/~t/REL606.fa.gz</code></pre>
<p>All programs have parameters that are associated with options from the program. You can read documentation here for <a href="http://www.gnu.org/software/wget/manual/wget.html">wget</a> and <a href="http://curl.haxx.se/docs/manual.html">curl</a>. A note: if you download software from Sourceforge, you'll want to get familiar with the curl -L command with the redirect to a file &quot;&gt;&quot; which you'll learn later.</p>
<h2 id="workign-with-the-ncbi-sra">Workign with the NCBI SRA</h2>
<p>The NCBI Sequence Read Archive is a large (&gt;3 quadrillion basepairs as of 2014) repository for next-generation sequence data. Like many NCBI database it is complex and mastering its use is greater than the scope of this lesson. Very often, as in the Lenski paper, there will be a direct link (perhaps in the supplemental information) to where on the SRA the dataset can be found. The link from the Lenski paper is: http://www.ncbi.nlm.nih.gov/sra?term=SRA026813</p>
<h2 id="exercises">Exercises</h2>
<blockquote>
<p>Sections A and B of these exercises take place on your own laptop - no need to open the shell or connect to your remote computer yet!</p>
</blockquote>
<h4 id="a.-locate-the-run-accessor-for-the-lenski-dataset-on-the-sra">A. Locate the Run Accessor for the Lenski Dataset on the SRA</h4>
<ol style="list-style-type: decimal">
<li>Access the Lenski dataset from the provided link: http://www.ncbi.nlm.nih.gov/sra?term=SRA026813 &gt; You will be presented with a page for the overall SRA accession SRA026813 - this is a collection of all the experimental data</li>
<li>Click on the first entry (<a href="http://www.ncbi.nlm.nih.gov/sra/SRX040669%5Baccn%5D">ZDB30</a>); this will take you to an a page for an SRX (Sequence Read eXperiment). Take a few minutes to examine some of the descriptions on the page</li>
<li>Click on the <a href="http://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP004752">Run Selector Link</a>; this is a description of all of the NGS datasets related to the experiment.</li>
<li>Go to the top of the page and <strong>In the 'Total row'</strong> you will see there are 37 runs, 10.15Gb data, and 16.45 Gbases of data. Click the 'RunInfo Table' button. &gt; We are not downloading any actual sequence data here! This is only a text file that fully describes the entire dataset</li>
</ol>
<p>You should now have a file called <strong>SraRunTable.txt</strong></p>
<h4 id="b.-review-the-sraruntable-in-a-spreadsheet-program">B. Review the SraRunTable in a spreadsheet program</h4>
<ol style="list-style-type: decimal">
<li>Using your choice of spreadsheet program open the <strong>SraRunTable.txt</strong> file. If prompted this is a tab-delimited file.</li>
</ol>
<p>Taking a quick glance at the file, you should be able to answer the following questions:</p>
<ol style="list-style-type: decimal">
<li>What strain of <em>E. coli</em> was used in this experiment?</li>
<li>What was the sequencing platform used for this experiment?</li>
<li>What samples in the experiment contain <a href="http://www.illumina.com/technology/next-generation-sequencing/paired-end-sequencing_assay.html">paired end</a> sequencing data?</li>
</ol>
<p>After answering the question, you should avoid saving this file; we don't want to make any changes. If you were to save this file, make sure you save it as a plain <strong>.txt</strong> file.</p>
<h3 id="c.-looking-at-a-text-file-at-the-command-line">C. Looking at a text file at the command line</h3>
<p>We are going to be doing our formal shell lesson next. However, we can already play with a few simple commands that have you thinking about how the shell can be useful tool in examining your dataset</p>
<blockquote>
<p>At this point, please verify that you are working from the the ami-6516b30e image where we have prestaged all the data for the rest of the lessons</p>
</blockquote>
<p>During the shell lesson we will go much more step-by-step, building our way through every command before we use it. For this first command go ahead any copy and paste if you are just getting use to the shell. The other commands are short enough to type and follow.</p>
<ol style="list-style-type: decimal">
<li><p>Follow you instructor's direction to navigate to the sample SraRunTable.txt file that has be placed on your remote computer</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">cd</span> ~/dc_sample_data/sra_metadata</code></pre></div></li>
<li><p>Display the contents of the SraRunTable.txt on your screen</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">cat</span> SraRunTable.txt</code></pre></div></li>
<li><p>That's probably not too pretty, so lets view it in a little more organized a format</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">column</span> -t SraRunTable.txt</code></pre></div></li>
<li><p>In this case we are using the 'column' program to make the print out more pretty.</p></li>
</ol>
<p>In the next lesson on the shell, we will be able to find out how to do a lot more with the shell, including what else the commands we have done can do.</p>
<h2 id="data-description">Data Description</h2>
<p>For the purposes of this workshop, we will be working with 6 of the fastq reads used in this experiment.</p>
<table>
<thead>
<tr class="header">
<th align="left">SRA Run Number</th>
<th align="left">Clone</th>
<th align="left">Generation</th>
<th align="left">Clade</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">SRR098028</td>
<td align="left">REL1166A</td>
<td align="left">2,000</td>
<td align="left">?</td>
</tr>
<tr class="even">
<td align="left">SRR098281</td>
<td align="left">ZDB409</td>
<td align="left">5,000</td>
<td align="left">?</td>
</tr>
<tr class="odd">
<td align="left">SRR098283</td>
<td align="left">ZDB446</td>
<td align="left">15,000</td>
<td align="left">UC</td>
</tr>
<tr class="even">
<td align="left">SRR097977</td>
<td align="left">CZB152</td>
<td align="left">33,000</td>
<td align="left">Cit+</td>
</tr>
<tr class="odd">
<td align="left">SRR098026</td>
<td align="left">CZB154</td>
<td align="left">33,000</td>
<td align="left">Cit+</td>
</tr>
<tr class="even">
<td align="left">SRR098027</td>
<td align="left">CZB199</td>
<td align="left">33,000</td>
<td align="left">C1</td>
</tr>
</tbody>
</table>
<h2 id="where-to-learn-more">Where to learn more</h2>
<h4 id="about-the-sequence-read-archive">About the Sequence Read Archive</h4>
<ul>
<li>You can learn more about the SRA by reading the <a href="http://www.ncbi.nlm.nih.gov/Traces/sra/">SRA Documentation</a></li>
<li>The best way to transfer a large SRA dataset is by using the <a href="http://www.ncbi.nlm.nih.gov/Traces/sra/?view=toolkit_doc">SRA Toolkit</a></li>
</ul>
</body>
</html>
