<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
  </style>
</head>
<body>
<h1 id="lesson-qc-of-sequence-read-data">Lesson QC of Sequence Read Data</h1>
<h1 id="quality-control-of-ngs-data">Quality Control of NGS Data</h1>
<h2 id="learning-objectives">Learning Objectives:</h2>
<h4 id="whats-the-goal-for-this-lesson">What's the goal for this lesson?</h4>
<ul>
<li>Understand how the FastQ format encodes quality</li>
<li>Be able to evaluate a FastQC report</li>
<li>Use Trimmommatic to clean FastQ reads</li>
<li>Use a For loop to automate operations on multiple files</li>
</ul>
<h2 id="details-on-the-fastq-format">Details on the FASTQ format</h2>
<p>Although it looks complicated (and maybe it is), its easy to understand the <a href="https://en.wikipedia.org/wiki/FASTQ_format">fastq</a> format with a little decoding. Some rules about the format include...</p>
<table>
<thead>
<tr class="header">
<th align="left">Line</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="left">Always begins with '@' and then information about the read</td>
</tr>
<tr class="even">
<td align="left">2</td>
<td align="left">The actual DNA sequence</td>
</tr>
<tr class="odd">
<td align="left">3</td>
<td align="left">Always begins with a '+' and sometimes the same info in line 1</td>
</tr>
<tr class="even">
<td align="left">4</td>
<td align="left">Has a string of characters which represent the quality scores; must have same number of characters as line 2</td>
</tr>
</tbody>
</table>
<p>so for example in our data set, one complete read is:</p>
<pre><code>$ head -n4 SRR098281.fastq 
@SRR098281.1 HWUSI-EAS1599_1:2:1:0:318 length=35
CNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+SRR098281.1 HWUSI-EAS1599_1:2:1:0:318 length=35
#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!</code></pre>
<p>This is a pretty bad read.</p>
<p>Notice that line 4 is:</p>
<pre><code>#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!</code></pre>
<p>As mentioned above, line 4 is a encoding of the quality. In this case, the code is the <a href="https://en.wikipedia.org/wiki/ASCII#ASCII_printable_code_chart">ASCII</a> character table. According to the chart a '#' has the value 35 and '!' has the value 33 - <strong>But these values are not actually the quality scores!</strong> There are actually several historical differences in how Illumina and other players have encoded the scores. Heres the chart from wikipedia:</p>
<pre><code>  SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS.....................................................
  ..........................XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX......................
  ...............................IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII......................
  .................................JJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJ......................
  LLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLL....................................................
  !&quot;#$%&amp;&#39;()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~
  |                         |    |        |                              |                     |
 33                        59   64       73                            104                   126
  0........................26...31.......40                                
                           -5....0........9.............................40 
                                 0........9.............................40 
                                    3.....9.............................40 
  0.2......................26...31........41                              

 S - Sanger        Phred+33,  raw reads typically (0, 40)
 X - Solexa        Solexa+64, raw reads typically (-5, 40)
 I - Illumina 1.3+ Phred+64,  raw reads typically (0, 40)
 J - Illumina 1.5+ Phred+64,  raw reads typically (3, 40)
     with 0=unused, 1=unused, 2=Read Segment Quality Control Indicator (bold) 
     (Note: See discussion above).
 L - Illumina 1.8+ Phred+33,  raw reads typically (0, 41)
 ```
 So using the Illumina 1.8 encouding, which is what you will mostly see from now on, our first c is called with a Phred score of 0 and our Ns are called with a score of 2. Read quality is assessed using the Phred Quality Score.  This score is logarithmically based and the score values can be interpreted as follows:

|Phred Quality Score |Probability of incorrect base call |Base call accuracy|
|:-------------------|:---------------------------------:|-----------------:|
|10 |1 in 10 |  90%|
|20 |1 in 100|  99%|
|30 |1 in 1000| 99.9%|
|40 |1 in 10,000|   99.99%|
|50 |1 in 100,000|  99.999%|
|60 |1 in 1,000,000|    99.9999%|

## FastQC
FastQC (http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) provides a simple way to do some quality control checks on raw sequence data coming from high throughput sequencing pipelines. It provides a modular set of analyses which you can use to give a quick impression of whether your data has any problems of which you should be aware before doing any further analysis.

The main functions of FastQC are
* Import of data from BAM, SAM or FastQ files (any variant)
* Providing a quick overview to tell you in which areas there may be problems
* Summary graphs and tables to quickly assess your data
* Export of results to an HTML based permanent report
* Offline operation to allow automated generation of reports without running the interactive application


## Running FASTQC
### A. Stage your data

1. Create a working directory for your analysis
   
    ```bash
    $ cd
    # this command takes us to the home directory
    
    $ mkdir dc_workshop
    ```
2. Create three three subdirectories

   ```bash
    mkdir dc_workshop/data
    mkdir dc_workshop/docs
    mkdir dc_workshop/results</code></pre>
<blockquote>
<p>The sample data we will be working with is in a hidden directory (placing a '.' in front of a directory name hides the directory. In the next step we will move some of those hidden files into our new dirctories to start our project. 3. Move our sample data to our working (home) directory</p>
</blockquote>
<p><code>bash  $ mv ~/.dc_sampledata_lite/untrimmed_fastq/ ~/dc_workshop/data/</code></p>
<h3 id="b.-run-fastqc">B. Run FastQC</h3>
<ol style="list-style-type: decimal">
<li><p>Navigate to the initial fastq dataset</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">cd</span> ~/dc_workshop/data/untrimmed_fastq/</code></pre></div>
To run the fastqc program, we call it from its location in <code>~/FastQC</code>. fastqc will accept multiple file names as input, so we can use the *.fastq wildcard.</li>
<li><p>Run FastQC on all fastq files in the directory</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">~/FastQC/fastqc</span> *.fastq</code></pre></div>
<p>Now, let's create a home for our results</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">mkdir</span> ~/dc_workshop/results/fastqc_untrimmed_reads</code></pre></div></li>
<li><p>Next, move the files there (recall, we are still in <code>~/dc_workshop/data/untrimmed_fastq/</code>) <code>bash  $ mv *.zip ~/dc_workshop/results/fastqc_untrimmed_reads/ $ mv *.html ~/dc_workshop/results/fastqc_untrimmed_reads/</code> ###C. Results</p></li>
</ol>
<p>Lets examine the results in detail</p>
<ol style="list-style-type: decimal">
<li>Navigate to the results and view the directory contents</li>
</ol>
<p><code>bash $ cd ~/dc_workshop/results/fastqc_untrimmed_reads/ $ ls</code></p>
<blockquote>
<p>The zip files need to be unpacked with the 'unzip' program.<br />
2. Use unzip to unzip the FastQC results: ```bash $ unzip *.zip</p>
</blockquote>
<pre><code>Did it work? No, because &#39;unzip&#39; expects to get only one zip file.  Welcome to the real world. We *could* do each file, one by one, but what if we have 500 files?  There is a smarter way. We can save time by using a simple shell &#39;for loop&#39; to iterate through the list of files in *.zip. After you type the first line, you will get a special &#39;&gt;&#39; prompt to type next next lines. You start with &#39;do&#39;, then enter your commands, then end with &#39;done&#39; to execute the loop.
3. Build a ``for`` loop to unzip the files

   ```bash 
$ for zip in *.zip
&gt; do
&gt; unzip $zip
&gt; done</code></pre>
<p>Note that, in the first line, we create a variable named 'zip'. After that, we call that variable with the syntax $zip. $zip is assigned the value of each item (file) in the list *.zip, once for each iteration of the loop.</p>
<p>This loop is basically a simple program. When it runs, it will run unzip once for each file (whose name is stored in the $zip variable). The contents of each file will be unpacked into a separate directory by the unzip program.</p>
<p>The for loop is interpreted as a multipart command. If you press the up arrow on your keyboard to recall the command, it will be shown like so: <code>bash     for zip in *.zip; do echo File $zip; unzip $zip; done</code></p>
<p>When you check your history later, it will help your remember what you did!</p>
<h3 id="d.-document-your-work">D. Document your work</h3>
<p>To save a record, let's cat all fastqc summary.txts into one full_report.txt and move this to <code>~/dc_workshop/docs</code>. You can use wildcards in paths as well as file names. Do you remember how we said 'cat' is really meant for concatenating text files?</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">cat</span> */summary.txt <span class="kw">&gt;</span> ~/dc_workshop/docs/fastqc_summaries.txt</code></pre></div>
<h2 id="how-to-clean-reads-using-trimmomatic">How to clean reads using <em>Trimmomatic</em></h2>
<h3 id="a-detailed-explanation-of-features">A detailed explanation of features</h3>
<p>Once we have an idea of the quality of our raw data, it is time to trim away adapters and filter out poor quality score reads. To accomplish this task we will use <em>Trimmomatic</em> <a href="http://www.usadellab.org/cms/?page=trimmomatic" class="uri">http://www.usadellab.org/cms/?page=trimmomatic</a>.</p>
<p><em>Trimmomatic</em> is a java based program that can remove sequencer specific reads and nucleotides that fall below a certain threshold. <em>Trimmomatic</em> can be multithreaded to run quickly.</p>
<p>Because <em>Trimmomatic</em> is java based, it is run using the command:</p>
<p><strong><em>java jar trimmomatic-0.32.jar</em></strong></p>
<p>What follows this are the specific commands that tells the program exactly how you want it to operate. <em>Trimmomatic</em> has a variety of options and parameters:</p>
<ul>
<li><strong><em>-threds</em></strong> How many processors do you want <em>Trimmomatic</em> to run with?</li>
<li><strong><em>SE</em></strong> or <strong><em>PE</em></strong> Single End or Paired End reads?</li>
<li><strong><em>-phred33</em></strong> or <strong><em>-phred64</em></strong> Which quality score do your reads have?</li>
<li><strong><em>SLIDINGWINDOW</em></strong> Perform sliding window trimming, cutting once the average quality within the window falls below a threshold.</li>
<li><strong><em>LEADING</em></strong> Cut bases off the start of a read, if below a threshold quality.</li>
<li><strong><em>TRAILING</em></strong> Cut bases off the end of a read, if below a threshold quality.</li>
<li><strong><em>CROP</em></strong> Cut the read to a specified length.</li>
<li><strong><em>HEADCROP</em></strong> Cut the specified number of bases from the start of the read.</li>
<li><strong><em>MINLEN</em></strong> Drop an entire read if it is below a specified length.</li>
<li><strong><em>TOPHRED33</em></strong> Convert quality scores to Phred-33.</li>
<li><strong><em>TOPHRED64</em></strong> Convert quality scores to Phred-64.</li>
</ul>
<p>A generic command for <em>Trimmomatic</em> looks like this:</p>
<p><strong>java jar trimmomatic-0.32.jar SE -thr</strong></p>
<p>A complete command for <em>Trimmomatic</em> will look something like this:</p>
<p><strong>java jar trimmomatic-0.32.jar SE -threads 4 -phred64 SRR_0156.fastq SRR_1056_trimmed.fastq ILLUMINACLIP:SRR_adapters.fa SLIDINGWINDOW:4:20</strong></p>
<p>This command tells <em>Trimmomatic</em> to run on a Single End file (<code>SRR_0156.fastq</code>, in this case), the output file will be called <code>SRR_0156_trimmed.fastq</code>, there is a file with Illumina adapters called <code>SRR_adapters.fa</code>, and we are using a sliding window of size 4 that will remove those bases if their phred score is below 20.</p>
<h2 id="exercise---running-trimmomatic">Exercise - Running Trimmomatic</h2>
<ol style="list-style-type: decimal">
<li>Go to the untrimmed fastq data location:</li>
</ol>
<p><code>bash $ cd /home/dcuser/dc_workshop/data/untrimmed_fastq</code></p>
<p>The command line incantation for trimmomatic is more complicated. This is where what you have been learning about accessing your command line history will start to become important.</p>
<p>The general form of the command is:</p>
<p><code>bash java -jar ~/Trimmomatic-0.32/trimmomatic-0.32.jar inputfile outputfile OPTION:VALUE...</code><br />
'java -jar' calls the Java program, which is needed to run trimmomargumentstic, which lived in a 'jar' file (trimmomatic-0.32.jar), a special kind of java archive that is often used for programs written in the Java programing language. If you see a new program that ends in '.jar', you will know it is a java program that is executed 'java -jar program name'. The 'SE' argument is a keyword that specifies we are working with single-end reads.</p>
<p>The next two arguments are input file and output file names. These are then followed by a series of options. The specifics of how options are passed to a program are different depending on the program. You will always have to read the manual of a new program to learn which way it expects its command-line arguments to be composed.</p>
<p>So, for the single fastq input file 'SRR098283.fastq', the command would be: ```bash $ java -jar /home/dcuser/Trimmomatic-0.32/trimmomatic-0.32.jar SE SRR098283.fastq<br />
 SRR098283.fastq_trim.fastq SLIDINGWINDOW:4:20 MINLEN:20</p>
<pre><code>TrimmomaticSE: Started with arguments: SRR098283.fastq SRR098283.fastq_trim.fastq SLIDINGWINDOW:4:20 MINLEN:20
Automatically using 2 threads
Quality encoding detected as phred33
Input Reads: 21564058 Surviving: 17030985 (78.98%) Dropped: 4533073 (21.02%)
TrimmomaticSE: Completed successfully</code></pre>
<pre><code>So that worked and we have a new fastq file.

   ```bash
    $ ls SRR098283*
    SRR098283.fastq  SRR098283.fastq_trim.fastq</code></pre>
<p>Now we know how to run trimmomatic but there is some good news and bad news.<br />
One should always ask for the bad news first. Trimmomatic only operates on one input file at a time and we have more than one input file. The good news? We already know how to use a for loop to deal with this situation.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">for</span> <span class="kw">infile</span> in *.fastq
    <span class="kw">&gt;do</span>
    <span class="kw">&gt;</span><span class="ot">outfile=$infile</span>\<span class="kw">_trim.fastq</span>
    <span class="kw">&gt;java</span> -jar ~/Trimmomatic-0.32/trimmomatic-0.32.jar SE <span class="ot">$infile</span> <span class="ot">$outfile</span> SLIDINGWINDOW:4:20 MINLEN:20
    <span class="kw">&gt;done</span></code></pre></div>
<p>Do you remember how the first specifies a variable that is assigned the value of each item in the list in turn? We can call it whatever we like. This time it is called infile. Note that the third line of this for loop is creating a second variable called outfile. We assign it the value of $infile with '_trim.fastq' appended to it. The '' escape character is used so the shell knows that whatever follows  is not part of the variable name $infile. There are no spaces before or after the '='.</p>
</body>
</html>
